cdc/owner/changefeed.go

```diff
@@ -0,0 +1,423 @@
+	c.state = state
+	c.feedStateManager.Tick(state)
+	if !c.feedStateManager.ShouldRunning() {
+		c.releaseResources()
+		return nil
+	}
+
+	checkpointTs := c.state.Info.GetCheckpointTs(c.state.Status)
+	if err := c.gcManager.CheckStaleCheckpointTs(ctx, checkpointTs); err != nil {
+		return errors.Trace(err)
+	}
+	if !c.preCheck(captures) {
+		return nil
+	}
+	if err := c.initialize(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	select {
+	case err := <-c.errCh:
+		return errors.Trace(err)
+	default:
+	}
+
+	c.sink.EmitCheckpointTs(ctx, checkpointTs)
+	barrierTs, err := c.handleBarrier(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	shouldUpdateState, err := c.scheduler.Tick(c.state, c.schema.AllPhysicalTables(), captures)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if shouldUpdateState {
+		c.updateStatus(barrierTs)
+	}
+	return nil
+}
+
+func (c *changefeed) initialize(ctx cdcContext.Context) error {
+	if c.initialized {
+		return nil
+	}
+	// empty the errCh
+LOOP:
+	for {
+		select {
+		case <-c.errCh:
+		default:
+			break LOOP
+		}
+	}
+	startTs := c.state.Info.GetCheckpointTs(c.state.Status)
+	log.Info("initialize changefeed", zap.String("changefeed", c.state.ID),
+		zap.Stringer("info", c.state.Info),
+		zap.Uint64("checkpoint ts", startTs))
+	failpoint.Inject("NewChangefeedNoRetryError", func() {
+		failpoint.Return(cerror.ErrStartTsBeforeGC.GenWithStackByArgs(startTs-300, startTs))
+	})
+
+	failpoint.Inject("NewChangefeedRetryError", func() {
+		failpoint.Return(errors.New("failpoint injected retriable error"))
+	})
+
+	if c.state.Info.Config.CheckGCSafePoint {
+		err := util.CheckSafetyOfStartTs(ctx, ctx.GlobalVars().PDClient, c.state.ID, startTs)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	c.barriers.Update(ddlJobBarrier, startTs)
+	c.barriers.Update(syncPointBarrier, startTs)
+	c.barriers.Update(finishBarrier, c.state.Info.GetTargetTs())
+	var err error
+	c.schema, err = newSchemaWrap4Owner(ctx.GlobalVars().KVStorage, startTs, c.state.Info.Config)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cancelCtx, cancel := cdcContext.WithCancel(ctx)
+	c.cancel = cancel
+	c.sink, err = c.newSink(cancelCtx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = c.sink.Initialize(cancelCtx, c.schema.SinkTableInfos())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	c.ddlPuller, err = c.newDDLPuller(cancelCtx, startTs)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		ctx.Throw(c.ddlPuller.Run(cancelCtx))
+	}()
+
+	// init metrics
+	c.metricsChangefeedCheckpointTsGauge = changefeedCheckpointTsGauge.WithLabelValues(c.id)
+	c.metricsChangefeedCheckpointTsLagGauge = changefeedCheckpointTsLagGauge.WithLabelValues(c.id)
+	c.initialized = true
+	return nil
+}
+
+func (c *changefeed) releaseResources() {
+	if !c.initialized {
```

Comment:
How is it guaranteed that there will be no resources that need releasing, if the changefeed has never been initialized?

Reply:
Unfortunately, it can only be guaranteed by hand.
The programmer should make sure that instances create in `newChangefeed` needless to release and make sure all the instances which need to be released are closed in `releaseResources` function.